pkg/controller/servicecatalog/provisionedproduct/setup.go

```diff
@@ -0,0 +1,268 @@
+	if pointer.StringDeref(cr.Status.AtProvider.Status, "") == string(svcapitypes.ProvisionedProductStatus_SDK_UNDER_CHANGE) {
+		return true, nil
+	}
+	obj.TerminateToken = aws.String(genIdempotencyToken())
+	obj.ProvisionedProductName = aws.String(meta.GetExternalName(cr))
+	return false, nil
+}
+
+func (c *custom) preUpdate(_ context.Context, cr *svcapitypes.ProvisionedProduct, input *svcsdk.UpdateProvisionedProductInput) error {
+	input.UpdateToken = aws.String(genIdempotencyToken())
+	input.ProvisionedProductName = aws.String(meta.GetExternalName(cr))
+	return nil
+}
+
+func provisioningParamsAreNotChanged(cfStackParams []cfsdkv2types.Parameter, currentParams []*svcapitypes.ProvisioningParameter) bool {
+	if len(cfStackParams) != len(currentParams) {
+		return false
+	}
+
+	cfStackKeyValue := make(map[string]string)
+	for _, v := range cfStackParams {
+		cfStackKeyValue[*v.ParameterKey] = *v.ParameterValue
+	}
+
+	for _, v := range currentParams {
+		if cfv, ok := cfStackKeyValue[*v.Key]; ok && cfv == *v.Value {
+			continue
+		} else {
+			return false
+		}
+	}
+	return true
+}
+
+func (c *custom) productOrArtifactAreNotChanged(ds *svcapitypes.ProvisionedProductParameters, resp *svcsdk.ProvisionedProductDetail) (bool, error) {
+	// ProvisioningArtifactID and ProvisioningArtifactName are mutual exclusive params, the same about ProductID and ProductName
+	// But if describe a provisioned product aws api will return only IDs, so it's impossible to compare names with ids
+	// Conditional statement below works only if desired state includes ProvisioningArtifactID and ProductID
+	if ds.ProvisioningArtifactID != nil && ds.ProductID != nil &&
+		(*ds.ProvisioningArtifactID != *resp.ProvisioningArtifactId ||
+			*ds.ProductID != *resp.ProductId) {
+		return false, nil
+		// In case if desired state includes not only IDs provider runs func `getArtifactID`, which produces
+		// additional request to aws api and retrieves an artifact id based on ProductId/ProductName and ProvisioningArtifactId/ProvisioningArtifactName
+		// for further comparison with artifact id in the current state
+	} else if ds.ProvisioningArtifactName != nil || ds.ProductName != nil {
+		desiredArtifactID, err := c.getArtifactID(ds)
+		if err != nil {
+			return false, err
+		}
+		if desiredArtifactID != *resp.ProvisioningArtifactId {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+func (c *custom) getArtifactID(crParams *svcapitypes.ProvisionedProductParameters) (string, error) {
+	input := svcsdk.DescribeProvisioningArtifactInput{
+		ProductId:                crParams.ProductID,
+		ProvisioningArtifactId:   crParams.ProvisioningArtifactID,
+		ProductName:              crParams.ProductName,
+		ProvisioningArtifactName: crParams.ProvisioningArtifactName,
+	}
+	output, err := c.client.DescribeProvisioningArtifact(&input)
+	if err != nil {
+		return "", err
+	}
+	return *output.ProvisioningArtifactDetail.Id, nil
```

Comment:
You might run into a panic when dereferencing a nil value. Better to check if it is nil before returning it.

Reply:
Aws sdk v1 says it can't by nil, or you recommend to deref it anyway, just in case?